pkg/server/server.go

```diff
@@ -48,37 +55,77 @@ func New(ctx context.Context, cl client.Client) *InventoryServer {
 		authenticators: []authenticator{
 			tpm.New(ctx, cl),
 		},
+		registrationCache: &registrationCache{
+			Mutex:         &sync.Mutex{},
+			registrations: make(map[string]registrationData),
+		},
 	}
 
 	return server
 }
 
-func (i *InventoryServer) getRancherCACert() string {
-	setting := &managementv3.Setting{}
-	if err := i.Get(i, types.NamespacedName{Name: "cacerts"}, setting); err != nil {
-		logrus.Errorf("Error getting cacerts setting: %s", err.Error())
-		return ""
+func (i *InventoryServer) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	escapedPath := sanitizeUserInput(req.URL.Path)
+
+	logrus.Infof("Incoming HTTP request for %s", escapedPath)
+
+	// expect /elemental/{api}...
+	splittedPath := strings.Split(escapedPath, "/")
+	if len(splittedPath) < 3 {
+		http.Error(resp, "", http.StatusNotFound)
+		return
 	}
+	// drop 'elemental'
+	splittedPath = splittedPath[2:]
+	api := splittedPath[0]
 
-	if setting.Value == "" {
-		if err := i.Get(i, types.NamespacedName{Name: "internal-cacerts"}, setting); err != nil {
-			logrus.Errorf("Error getting internal-cacerts setting: %s", err.Error())
-			return ""
+	switch api {
+	case "build-image":
+		if err := i.apiBuildImage(resp, req); err != nil {
+			logrus.Errorf("build-image: %s", err.Error())
+			return
 		}
+	case "registration":
+		if err := i.apiRegistration(resp, req); err != nil {
+			logrus.Errorf("registration: %s", err.Error())
+			return
+		}
+	default:
+		logrus.Errorf("Unknown API: %s", api)
+		http.Error(resp, fmt.Sprintf("unknwon api: %s", api), http.StatusBadRequest)
+		return
 	}
-	return setting.Value
 }
 
-func (i *InventoryServer) getRancherServerURL() (string, error) {
-	setting := &managementv3.Setting{}
-	if err := i.Get(i, types.NamespacedName{Name: "server-url"}, setting); err != nil {
-		return "", fmt.Errorf("failed to get server url setting: %w", err)
+func sanitizeUserInput(data string) string {
+	escapedData := strings.Replace(data, "\n", "", -1)
+	escapedData = strings.Replace(escapedData, "\r", "", -1)
+	escapedData = html.EscapeString(escapedData)
+	return escapedData
+}
+
+func upgrade(resp http.ResponseWriter, req *http.Request) (*websocket.Conn, error) {
+	upgrader := websocket.Upgrader{
+		HandshakeTimeout: 5 * time.Second,
+		CheckOrigin:      func(r *http.Request) bool { return true },
 	}
 
-	if setting.Value == "" {
-		return "", fmt.Errorf("server-url is not set")
+	conn, err := upgrader.Upgrade(resp, req, nil)
+	if err != nil {
+		return nil, err
 	}
+	_ = conn.SetWriteDeadline(time.Now().Add(register.RegistrationDeadlineSeconds * time.Second))
```

Comment:
should we check for error here?

Reply:
*SetWriteDeadline* always return nil error but (the immediately following) *SetReadDeadline* may fail:
agreed, we should report the error but try to serve the websocket in any case.